naga/src/front/wgsl/lower/mod.rs

```diff
@@ -2450,52 +2463,218 @@ impl<'source, 'temp> Lowerer<'source, 'temp> {
 
                     crate::Expression::Derivative { axis, ctrl, expr }
                 } else if let Some(fun) = conv::map_standard_fun(function.name) {
-                    let expected = fun.argument_count() as _;
-                    let mut args = ctx.prepare_args(arguments, expected, span);
-
-                    let arg = self.expression(args.next()?, ctx)?;
-                    let arg1 = args
-                        .next()
-                        .map(|x| self.expression(x, ctx))
-                        .ok()
-                        .transpose()?;
-                    let arg2 = args
-                        .next()
-                        .map(|x| self.expression(x, ctx))
-                        .ok()
-                        .transpose()?;
-                    let arg3 = args
-                        .next()
-                        .map(|x| self.expression(x, ctx))
-                        .ok()
-                        .transpose()?;
+                    use crate::proc::OverloadSet as _;
+
+                    let fun_overloads = fun.overloads();
+                    let mut remaining_overloads = fun_overloads.clone();
+                    let min_arguments = remaining_overloads.min_arguments();
+                    let max_arguments = remaining_overloads.max_arguments();
+                    if arguments.len() < min_arguments || max_arguments < arguments.len() {
```

Comment:
style: `RangeFull::contains` seems like an easier-to-read alternative in the face of a moderately long expression-to-check like `arguments.len()`:

```suggestion
                    if !(min_arguments..=max_argument).contains(&arguments.len()) {
```

Reply:
Agreed that `Range` would be a nicer API.

However, in responding to earlier comments I changed this code to return `TooManyArguments` for that case, which gave us separate `if` statements for the too-few and too-many cases, meaning that `contains` isn't the right thing any more.

Reply:
I think you mean `Range::contains`, not `RangeFull::contains`, right?

Reply:
(Does returning a `Range<usize>` imply that every arity within the range is accepted?)

Reply:
Having tried this out, I'm not sure it's actually an improvement. Look at this:

```rust
                    let argument_counts = remaining_overloads.argument_counts();
                    if arguments.len() < argument_counts.start {
                        return Err(Box::new(Error::WrongArgumentCount {
                            span,
                            expected: argument_counts.start as u32..argument_counts.end as u32 + 1,
                            found: arguments.len() as u32,
                        }));
                    }
                    if arguments.len() >= argument_counts.end {
                        return Err(Box::new(Error::TooManyArguments {
                            function: fun.to_wgsl_for_diagnostics(),
                            call_span: span,
                            arg_span: ctx.ast_expressions.get_span(arguments[argument_counts.end - 1]),
                            max_arguments: (argument_counts.end - 1) as _,
                        }));
                    }
```

This seems worse:

- We're now talking about `counts.start` and `counts.end`, which are less legible than `min_arguments` and `max_arguments`.
- Since we want to use `Range`, not `RangeInclusive` with its distracting and irrelevant `Option`, the `end` needs to be one greater than the value we always actually want, so we get little fixups everywhere.

At first I thought it would be good, but having tried it out, I'm not fond of this change.

Reply:
If we're going to have separate errors be returned for each end of the bounds check, then I don't think the suggestion to use `RangeInclusive` (and yes, I meant to refer to that one, rather than `RangeFull` or even `Range 😅) is useful anymore.

Reply:
Closing, since the basis of this suggestion seems invalidated.